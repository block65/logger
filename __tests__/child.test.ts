import {
  afterAll,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { createCliTransformer, jsonTransformer, Level } from '../lib/index.js';

describe('Child Logger', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    jest.useFakeTimers({ now: new Date('2009-02-13T23:31:30.000Z') });
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  test('Basic', async () => {
    const { createLoggerWithWaitableMock } = await import('./helpers.js');

    const [logger, callback, errback] = createLoggerWithWaitableMock();

    const childLogger = logger.child({ helloChildLogger: 'hello!' });

    childLogger.fatal(new Error('fatal'));
    childLogger.error(new Error('error'));
    childLogger.info(new Error('info'));
    childLogger.warn(new Error('warn'));
    childLogger.debug(new Error('debug'));
    childLogger.trace(new Error('trace'));
    await expect(callback.waitUntilCalledTimes(5)).resolves.toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });

  test('with context + cli', async () => {
    const { createLoggerWithWaitableMock } = await import('./helpers.js');

    const [logger, callback, errback] = createLoggerWithWaitableMock({
      transformer: createCliTransformer(),
    });

    const childLogger = logger.child(
      { helloChildLogger: 'hello!' },
      {
        context: { name: 'CHILDLOGGER' },
      },
    );

    childLogger.fatal(new Error('fatal'));
    childLogger.error(new Error('error'));
    childLogger.info(new Error('info'));
    childLogger.warn(new Error('warn'));
    childLogger.debug(new Error('debug'));
    childLogger.trace(new Error('trace'));
    await expect(callback.waitUntilCalledTimes(6)).resolves.toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });

  test('level changes parent > child', async () => {
    const { createLoggerWithWaitableMock } = await import('./helpers.js');

    const [logger, callback, errback] = createLoggerWithWaitableMock({
      level: Level.Fatal,
      context: { parent: 'test' },
    });

    const childLogger = logger.child(
      { extraChildData: 'hello!' },
      {
        level: Level.Trace,
        context: {
          extraChildContext: 'hello!',
        },
      },
    );

    // logger.on('log', console.error);

    logger.fatal('fatal');
    childLogger.fatal('fatal');
    childLogger.fatal('fatal');
    childLogger.error('error');
    childLogger.info('info');
    childLogger.warn('warn');
    childLogger.debug('debug');
    childLogger.trace('trace');

    await expect(callback.waitUntilCalledTimes(2)).resolves.toMatchSnapshot();
    expect(errback).not.toBeCalled();
  }, 10000);

  test('level changes child > parent', async () => {
    const { createLoggerWithWaitableMock } = await import('./helpers.js');

    const [logger, callback, errback] = createLoggerWithWaitableMock({
      level: Level.Trace,
    });

    const childLogger = logger.child(
      { helloChildLogger: 'hello!' },
      {
        level: Level.Fatal,
      },
    );

    childLogger.fatal(new Error('fatal'));
    childLogger.error(new Error('error'));
    childLogger.info(new Error('info'));
    childLogger.warn(new Error('warn'));
    childLogger.debug(new Error('debug'));
    childLogger.trace(new Error('trace'));
    await expect(callback.waitUntilCalledTimes(1)).resolves.toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });

  test('child logger spam', async () => {
    const { createLoggerWithWaitableMock } = await import('./helpers.js');

    const [logger, callback, errback] = createLoggerWithWaitableMock({
      level: Level.Trace,
    });

    const warningMock = jest.fn();

    process.on('warning', (err) => {
      // eslint-disable-next-line no-console
      console.warn(err);
      warningMock(err);
    });

    const iterations = 500;

    const autoEndMock = jest.fn(() => {});

    // eslint-disable-next-line no-restricted-syntax
    for await (const childIndex of [...Array(iterations)].map(
      (_, idx) => idx,
    )) {
      const childLogger = logger.child({ childIndex });
      childLogger.info(new Error('hello'));
      childLogger.on('end', autoEndMock);
    }

    await logger.end();

    await callback.waitUntilCalledTimes(iterations);

    expect(errback).not.toBeCalled();
    expect(warningMock).not.toBeCalled();

    // make sure each child was ended
    expect(autoEndMock).toBeCalledTimes(iterations);
  });

  test('Lambda Context wrapper with child', async () => {
    process.env.AWS_LAMBDA_FUNCTION_VERSION = '$LATEST';

    const { withLambdaLoggerContextWrapper } = await import('../lib/lambda.js');

    const { createAutoConfiguredLoggerWithWaitableMock } = await import(
      './helpers.js'
    );

    const [logger, callback, errback] =
      createAutoConfiguredLoggerWithWaitableMock();

    logger.info('logger outside before');

    const childLogger = logger.child(
      {
        childData: '1',
      },
      {
        context: {
          childCtx: '1',
        },
      },
    );

    await withLambdaLoggerContextWrapper(
      childLogger,
      {
        awsRequestId: 'fake-request-id1',
        functionVersion: '999',
      },
      async () => {
        logger.info('logger inside');
        childLogger.info('childLogger inside');
      },
    );

    logger.info('logger outside after');

    await logger.end();

    await expect(callback.waitUntilCalledTimes(4)).resolves.toMatchSnapshot();

    expect(errback).not.toBeCalled();
  });

  test('Child message format', async () => {
    const { createLoggerWithWaitableMock } = await import('./helpers.js');

    const [logger, callback, errback] = createLoggerWithWaitableMock({
      transformer: jsonTransformer,
    });

    const childLogger = logger.child({ helloChildLogger: 'hello!' });

    childLogger.fatal(
      { err: new Error(), woo: 'yeah' },
      'Time is an %s',
      'illusion',
    );

    await expect(callback.waitUntilCalledTimes(1)).resolves.toMatchSnapshot();
    expect(errback).not.toBeCalled();
  });
});
