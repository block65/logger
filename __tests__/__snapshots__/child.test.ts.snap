// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Child Logger Basic 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fatal",
          "name": "Error",
          "stack": "Error: fatal
    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:28:23)",
        },
        "helloChildLogger": "hello!",
      },
      "err": [Error: fatal],
      "level": 60,
      "msg": "fatal",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "error",
          "name": "Error",
          "stack": "Error: error
    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:29:23)",
        },
        "helloChildLogger": "hello!",
      },
      "err": [Error: error],
      "level": 50,
      "msg": "error",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "info",
          "name": "Error",
          "stack": "Error: info
    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:30:22)",
        },
        "helloChildLogger": "hello!",
      },
      "err": [Error: info],
      "level": 30,
      "msg": "info",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "warn",
          "name": "Error",
          "stack": "Error: warn
    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:31:22)",
        },
        "helloChildLogger": "hello!",
      },
      "err": [Error: warn],
      "level": 40,
      "msg": "warn",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "debug",
          "name": "Error",
          "stack": "Error: debug
    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:32:23)",
        },
        "helloChildLogger": "hello!",
      },
      "err": [Error: debug],
      "level": 20,
      "msg": "debug",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "trace",
          "name": "Error",
          "stack": "Error: trace
    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:33:23)",
        },
        "helloChildLogger": "hello!",
      },
      "err": [Error: trace],
      "level": 10,
      "msg": "trace",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Child Logger Lambda Context wrapper with child 1`] = `
Array [
  Array [
    "2009-02-13T23:31:30.000Z		INFO	logger outside before	{\\"level\\":30}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		INFO	logger inside	{\\"level\\":30}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z	fake-request-id1	INFO	childLogger inside	{\\"level\\":30,\\"childData\\":\\"1\\",\\"ctx\\":{\\"context\\":{\\"functionVersion\\":\\"999\\"},\\"childCtx\\":\\"1\\"}}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		INFO	logger outside after	{\\"level\\":30}
",
  ],
]
`;

exports[`Child Logger level changes child > parent 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fatal",
          "name": "Error",
          "stack": "Error: fatal
    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:109:23)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
        },
        "helloChildLogger": "hello!",
      },
      "err": [Error: fatal],
      "level": 60,
      "msg": "fatal",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Child Logger level changes parent > child 1`] = `
Array [
  Array [
    Object {
      "ctx": Object {
        "parent": "test",
      },
      "level": 60,
      "msg": "fatal",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "ctx": Object {
        "extraChildContext": "hello!",
      },
      "data": Object {
        "extraChildData": "hello!",
      },
      "level": 60,
      "msg": "fatal",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "ctx": Object {
        "extraChildContext": "hello!",
      },
      "data": Object {
        "extraChildData": "hello!",
      },
      "level": 60,
      "msg": "fatal",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Child Logger with context + cli 1`] = `
Array [
  Array [
    "2009-02-13T23:31:30.000Z FATAL  (CHILDLOGGER) fatal { err: { message: 'fatal', name: 'Error', stack: 'Error: fatal\\\\n    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:52:23)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)' }, helloChildLogger: 'hello!' }
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z ERROR  (CHILDLOGGER) error { err: { message: 'error', name: 'Error', stack: 'Error: error\\\\n    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:53:23)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)' }, helloChildLogger: 'hello!' }
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z INFO   (CHILDLOGGER) info { err: { message: 'info', name: 'Error', stack: 'Error: info\\\\n    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:54:22)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)' }, helloChildLogger: 'hello!' }
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z WARN   (CHILDLOGGER) warn { err: { message: 'warn', name: 'Error', stack: 'Error: warn\\\\n    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:55:22)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)' }, helloChildLogger: 'hello!' }
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z DEBUG  (CHILDLOGGER) debug { err: { message: 'debug', name: 'Error', stack: 'Error: debug\\\\n    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:56:23)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)' }, helloChildLogger: 'hello!' }
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z TRACE  (CHILDLOGGER) trace { err: { message: 'trace', name: 'Error', stack: 'Error: trace\\\\n    at Object.<anonymous> (~/block65/logger/__tests__/child.test.ts:57:23)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)' }, helloChildLogger: 'hello!' }
",
  ],
]
`;
